parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -1093,10 +1099,38 @@ private ColumnChunkPageReadStore internalReadFilteredRowGroup(BlockMetaData bloc
         }
       }
     }
-    // actually read all the chunks
+    // Vectored IO up.
+
+    List<FileRange> ranges = new ArrayList<>();
     for (ConsecutivePartList consecutiveChunks : allParts) {
-      consecutiveChunks.readAll(f, builder);
+      ranges.add(FileRange.createFileRange(consecutiveChunks.offset, (int) consecutiveChunks.length));
+    }
+    LOG.warn("Doing vectored IO for ranges {}", ranges);
+    f.readVectored(ranges, ByteBuffer::allocate);
```

Comment:
Yes, vectored read allocates a single buffer per range. 
If it can grow more than a GB, then I guess we will run into memory problems.

Reply:
Well, I just went through the code of ConsecutivePartList#readAll() again. Yes, they are breaking the big range into smaller buffers but allocating all of them in one go only, so won't the memory issue still persists?

Also, if I do the change in readAll() like I have already done the commented readAllVectored(), we really won't be reducing the number of seek operations thus won't be getting the real benefits of vectored IO. It will just be like there is a big range to be fetched, we break into smaller ranges and fetch them parallelly. ( This is similar to PARQUET-2149 which you proposed and have already uploaded the PR :) ).

Reply:
So if  I have  something like this - 
 `range 1, delta 1, range2`
I will give you two ranges, `range1`, and `range2` separated by a gap of `delta1` bytes. If `delta1` is small enough, you will merge `range1` and `range2` and do a single scan. 
If `range1` and `range2` are very large so that the resulting range is even larger, do you still do a single scan, or do you split the large range into smaller ranges and tradeoff some seek cost for increased parallelism?

Reply:
> if `delta1` is small enough, you will merge `range1` and `range2` and do a single scan.
Yes.

> If range1 and range2 are very large so that the resulting range is even larger, do you still do a single scan, or do you split the large range into smaller ranges and tradeoff some seek cost for increased parallelism?

We don't split the ranges. If the ranges are big enough, they won't be merged.

Reply:
> We don't split the ranges. If the ranges are big enough, they won't be merged.

How big  is 'big enough' ? 
Perhaps, in your approach, you can break up a large `consecutiveParts` into 'big enough' sized ranges,  instead of one large range, to get the benefit of parallelism?

Reply:
Also, re memory allocation - as long as we are allocating the same amount of memory as the current code is allocating, we should be fine.

Reply:
https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/PositionedReadable.java#L105
This is the max size by default. Also this can be configured for S3 https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java#L1196

Reply:
Hmm. The largest you will get after a merge is 1 MB (default for S3). But if parquet gives provides an 8 MB range vectored io will not split it into 8 ranges of 1MB. The _smallest_ read a parquet file reader is likely to do is 1 MB (the default page size), so in effect we are never going to merge ranges. 
Either way, we have two goals - reduce the number of seeks and increase the parallelism -  which seem to be in conflict. If we increase the number of ranges (and consequently use smaller ranges) we get more seeks, and if we decrease the number of ranges we get less parallelism. 
Do you have any suggestions on what is a good compromise based on experiments with Hdfs/S3 etc?

Reply:
I haven't been able to do any benchmarks for Parquet yet. Also, we can always configure or even change the default later based on the outcome of the experiments we perform. 

We have only done the benchmarks for Hive in orc till now. Even for the hive, I want to run more benchmarks by configuring the minseek and maxsize with different values and figure out the best default values.

Reply:
I have a prototype that mimics what I'm thinking. Let me share that.

Reply:
@mukund-thakur take a look at the two most recent commits here - https://github.com/parthchandra/incubator-parquet-mr/commits/mergeScanRanges
The parallel reader takes a range and splits it equally across the thread pool (i.e there are as many streams as there are threads in the pool).

Reply:
Thanks for sharing. I actually saw those changes in your original PR as well. Don't you think it is similar to what I have done in the commented readVectored() implementation?
I guess we need to run some benchmarks. First, I need to see what type of ranges parquet generates for real-world/tpcds queries. Do you have that by any chance?

Reply:
> I actually saw those changes in your original PR as well.

You couldn't have ! :) My PR for async IO is rather different and a lot more complicated.

Reply:
> Don't you think it is similar to what I have done in the commented readVectored() implementation?
Yes, it is. However, the parallel reading will split a large range as opposed to what the vectored read will do.
> I guess we need to run some benchmarks. First, I need to see what type of ranges parquet generates for real-world/tpcds queries. Do you have that by any chance?
No I don't have the ranges. TPCDS is a large set of queries. Pick a scale factor and a couple of queries and I'll see what I can get for you.

Reply:
> You couldn't have ! :) My PR for async IO is rather different and a lot more complicated.

Oh yeah just revisited. You are right. :)

Reply:
`    2022-10-18 21:51:28,308 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[1581775,2178915), range[6700883,9475957), range[10023390,10426141), range[12211215,15766053), range[24603672,25148984)]
`

I got these ranges while running tpcds query22 in hive on the parquet files in s3 with a scale factor of 1000. We can try the same in spark. thanks

Reply:
Let me try to get the corresponding ranges for Spark.

Reply:
that would be great. thanks. 
Also, I reran the same with changes in readAllVectored and I can see that it breaks the big ranges into smaller ones if the size is greater than maxAllocationSize ( default 8 MB). For exmaple see the breaking in range[24928095,34504532]. note- this example is for query26.

So I guess changing in the above layer makes more sense. 

> 2022-10-20 21:36:02,758 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[5326394,7673015), range[24928095,34504532), range[36603729,37991198), range[44105694,56402697), range[86874390,88752497)]
2022-10-20 21:36:03,146 [WARN] [TezChild] |hadoop.ParquetFileReader|: Reading through the vectored API.[readAllVectored]
2022-10-20 21:36:03,147 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[5326394,7673015)]
2022-10-20 21:36:03,225 [WARN] [TezChild] |hadoop.ParquetFileReader|: Reading through the vectored API.[readAllVectored]
2022-10-20 21:36:03,225 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[24928095,33316703), range[33316703,34504532)]
2022-10-20 21:36:03,352 [WARN] [TezChild] |hadoop.ParquetFileReader|: Reading through the vectored API.[readAllVectored]
2022-10-20 21:36:03,352 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[36603729,37991198)]
2022-10-20 21:36:03,439 [WARN] [TezChild] |hadoop.ParquetFileReader|: Reading through the vectored API.[readAllVectored]
2022-10-20 21:36:03,439 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[44105694,52494302), range[52494302,56402697)]
2022-10-20 21:36:03,652 [WARN] [TezChild] |hadoop.ParquetFileReader|: Reading through the vectored API.[readAllVectored]
2022-10-20 21:36:03,652 [WARN] [TezChild] |hadoop.ParquetFileReader|: Doing vectored IO for ranges [range[86874390,88752497)]